detection_rules/integrations.py

```diff
@@ -0,0 +1,128 @@
+
+    manifest_file = gzip.open(manifest_file_path, "w+")
+    manifest_file_bytes = json.dumps(final_integration_manifests).encode("utf-8")
+    manifest_file.write(manifest_file_bytes)
+
+
+def find_least_compatible_version(package: str, integration: str,
+                                  current_stack_version: str, packages_manifest: dict) -> str:
+    """Finds least compatible version for specified integration based on stack version supplied."""
+    integration_manifests = packages_manifest[package]
+    least_compatible_version = None
+    compatible_versions = {}
+
+    def compare_versions(int_ver: str, pkg_ver: str) -> bool:
+        """Compares integration and package version"""
+        pkg_major, pkg_minor = Version(pkg_ver)
+        int_major, int_minor = Version(int_ver)[:2]
+
+        if int(int_major) < int(pkg_major) or int(pkg_major) > int(int_major):
+            return False
+        compatible = Version(int_ver) <= Version(pkg_ver)
+        return compatible
+
+    for ver, manifest in integration_manifests.items():
+        kibana_compat_vers = re.sub(r"\>|\<|\=|\^", "", manifest["conditions"]["kibana.version"])
+        kibana_compat_vers = kibana_compat_vers.split(" || ")
+        bool_checks = [compare_versions(kcv, current_stack_version) for kcv in kibana_compat_vers]
+        if any(bool_checks):
+            compatible_versions.setdefault(ver, True)
+
+    if len(list(compatible_versions.keys())) > 0:
+        compatible_version_list = list(compatible_versions.keys())
+        compatible_version_list.sort(key=Version)
+        least_compatible_version = compatible_version_list[0]
+    else:
+        raise Exception(f"no compatible version for integration {package}:{integration}")
+    return least_compatible_version
+
+
+def get_integration_manifests(repository, sha: str, pkg_path: str) -> list:
+    """Iterates over specified integrations from package-storage and combines manifests per version."""
+    integration = pkg_path.split("/")[-1]
+    versioned_packages = repository.get_dir_contents(pkg_path, ref=sha)
+    versions = [p.path.split("/")[-1] for p in versioned_packages]
+    versioned_packages_contents = list()
+    for v in versions:
+        contents = repository.get_dir_contents(f"{pkg_path}/{v}", ref=sha)
+        versioned_packages_contents.append(contents)
+
+    print(f"Processing {integration} - Versions: {versions}")
+    manifests = []
+    for content in versioned_packages_contents:
+        processing_version = content[0].path.split("/")[2]
+        manifest_content = [c for c in content if "manifest" in c.path]
+        if len(manifest_content) < 1:
+            raise Exception(f"manifest file does not exist for {integration}:{processing_version}")
+        path = manifest_content[0].path
+        manifest_content = yaml.safe_load(repository.get_contents(path, ref=sha).decoded_content.decode())
+        manifests.append(manifest_content)
+
+    return manifests
```

Comment:
Use Path and remove redundant code

```suggestion
    integration = pkg_path.name
    versioned_packages = repository.get_dir_contents(pkg_path, ref=sha)
    versions = [p.path.split("/")[-1] for p in versioned_packages]

    manifests = []
    for version in versions:
        contents = repository.get_dir_contents(f"{pkg_path}/{version}", ref=sha)
        print(f"Processing {integration} - Versions: {versions}")

        processing_version = contents[0].path.split("/")[2]
        manifest_content = [c for c in contents if "manifest" in c.path]

        if len(manifest_content) < 1:
            raise Exception(f"manifest file does not exist for {integration}:{processing_version}")

        path = manifest_content[0].path
        manifest_content = yaml.safe_load(repository.get_contents(path, ref=sha).decoded_content.decode())
        manifests.append(manifest_content)

    return manifests
```